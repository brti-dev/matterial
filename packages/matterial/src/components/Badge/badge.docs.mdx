---
name: Badge
description: A small emblem to attach to avatars, icons, and other elements togenerate notifications.
---

## Usage

```jsx
import { Badge } from 'matterial'

function App() {
  return (
    <Badge content={99}>
      <Icon.Notification />
    </Badge>
  )
}
```

## Basic Badge

<Badge content={1}>
  <IconButton variant="contained" color="primary">
    <Icon.Notification />
  </IconButton>
</Badge>

```jsx
<Badge content={1}>
  <Icon.Notification />
</Badge>
```

## Colored Badge

<Container row>
  {[...COLORS, 'hotpink'].map(color => (
    <Badge key={color} content={99} color={color}>
      <Avatar alt={color}>{color.substring(0, 1).toUpperCase()}</Avatar>
    </Badge>
  ))}
</Container>

## Dot Badge

The `dot` variant changes a badge into a small dot. This can
be used as a notification that something has changed without giving a
count.

<Badge variant="dot" color="secondary">
  <IconButton variant="contained">
    <Icon.Notification />
  </IconButton>
</Badge>

```jsx
<Badge variant="dot" color="secondary">
  <Icon.Notification />
</Badge>
```

## Maximum Content

Use the <code>max</code> prop to specify the maximum number in the
badge.

<Badge color="red" max={99} content={100}>
  <IconButton variant="contained">
    <Icon.Notification />
  </IconButton>
</Badge>

```jsx
<Badge color="red" max={99} content={100}>
  <Icon.Notification />
</Badge>
```

## Badge Visibility

The badge automatically hides when <code>content</code> is zero. You can
override this with the <code>showZero</code> prop.

<Container>
  <Badge color="red" content={0}>
    <IconButton variant="contained">
      <Icon.Notification />
    </IconButton>
  </Badge>
  <Badge color="red" content={0} showZero>
    <IconButton variant="contained">
      <Icon.Notification />
    </IconButton>
  </Badge>
</Container>

```jsx
<Badge color="red" content={0}><Icon.Notification /></Badge>
<Badge color="red" content={0} showZero><Icon.Notification /></Badge>
```
