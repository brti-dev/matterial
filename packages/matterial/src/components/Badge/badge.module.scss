@use '../../styles/theme' as *;

.container {
  position: relative;
  display: inline;
}

.content {
  position: absolute;
  z-index: 1;
  top: 0;
  right: 0;
  display: flex;
  flex-flow: row wrap;
  place-content: center;
  align-items: center;
  box-sizing: border-box;
  font-weight: 500;
  font-size: 0.5rem;
  color: var(--color-font);
  min-width: 20px;
  padding: 0px 6px;
  height: 20px;
  border-radius: 10px;
  background-color: var(--color-background);
  box-shadow: 0 0 0 2px var(--color-body);
  transition: transform 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transform: scale(1) translate(50%, -50%);
  transform-origin: 100% 0%;
}

.variant--dot {
  width: 10px;
  height: 10px;
  padding: 0;
  min-width: auto;
  top: 3px;
  right: 3px;
}

// The badge content is a React Component rather than a string or number
.componentAsBadge {
  padding: 0;
}
