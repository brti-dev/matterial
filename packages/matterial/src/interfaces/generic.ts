/**
 * Remove properties `K` from `T`.
 * Distributive for union types.
 */
export type DistributiveOmit<T, K extends keyof any> = T extends any
  ? Omit<T, K>
  : never
