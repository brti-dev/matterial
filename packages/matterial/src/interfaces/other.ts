export type DateTimeType = Date | string | number
