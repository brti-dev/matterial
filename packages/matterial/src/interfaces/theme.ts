export type Severity = 'error' | 'warning' | 'info' | 'success'

export type Color =
  | Severity
  | 'default'
  | 'primary'
  | 'secondary'
  | 'red'
  | 'green'
  | 'dark'
  | 'light'

export type Variant = 'default' | 'contained' | 'outlined'
