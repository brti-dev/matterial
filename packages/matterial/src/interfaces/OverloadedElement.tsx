import { DistributiveOmit } from './generic'

/**
 * Elements that accept no children
 */
type VoidElement =
  | 'area'
  | 'base'
  | 'br'
  | 'col'
  | 'hr'
  | 'img'
  | 'input'
  | 'link'
  | 'meta'
  | 'param'
  | 'command'
  | 'keygen'
  | 'source'

type OmitChildrenFromVoid<C extends React.ElementType> = C extends VoidElement
  ? Omit<React.ComponentPropsWithRef<C>, 'children'>
  : React.ComponentPropsWithRef<C>

export interface OverloadedElement<P> {
  <C extends React.ElementType>(
    props: { as: C } & P & DistributiveOmit<OmitChildrenFromVoid<C>, keyof P>
  ): JSX.Element
  (
    props: P & DistributiveOmit<React.ComponentPropsWithRef<'div'>, keyof P>
  ): JSX.Element
}

export type OverloadedElementProps = {
  as?: string | React.ElementType
}
