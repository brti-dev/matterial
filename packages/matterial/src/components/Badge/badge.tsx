import React from 'react'

import { Color } from '../../interfaces/theme'
import classes from './badge.module.scss'

type BadgePropsBase = {
  children: React.ReactNode
  className?: string
  color?: Color
  max?: number
  showZero?: boolean
  size?: 'small' | 'medium' | 'large' | number
  variant?: 'default' | 'dot' | string
} & React.ComponentPropsWithoutRef<'span'>
type BadgePropsContent = BadgePropsBase & {
  content: string | number | null | React.ReactElement
}
type BadgePropsDot = BadgePropsBase & {
  content?: string | number | null | React.ReactElement
  variant: 'dot'
}

export type BadgeProps = BadgePropsContent | BadgePropsDot

export function Badge(props: BadgeProps) {
  const {
    children,
    className,
    color = 'default',
    content,
    max,
    showZero = false,
    size = 'medium',
    variant = 'default',
    ...rest
  } = props

  const classNames = [
    'badge',
    classes.content,
    'variant--contained', // Access global colors
    `color--${color}`,
    'no-hover',
    classes[`variant--${variant}`],
    classes[`size--${size}`],
    className && className,
  ]

  let contentOutput = content
  if (typeof content === 'number' && max && max < content) {
    contentOutput = `${max}+`
  }

  let hidden = false
  if (content === 0 && !showZero) {
    classNames.push('visually-hidden')
    hidden = true
  }

  if (React.isValidElement(content)) {
    classNames.push(classes.componentAsBadge)
  }

  return (
    <span className={classes.container}>
      {children}
      <span className={classNames.join(' ')} {...rest} hidden={hidden}>
        {contentOutput}
      </span>
    </span>
  )
}
