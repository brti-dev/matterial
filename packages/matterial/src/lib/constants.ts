import { Color } from 'interfaces/theme'

export const COLORS: Color[] = [
  'primary',
  'secondary',
  'error',
  'warning',
  'info',
  'success',
  'dark',
  'light',
]
