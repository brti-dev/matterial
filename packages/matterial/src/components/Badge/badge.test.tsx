import '@testing-library/jest-dom'
import renderer from 'react-test-renderer'

import { render } from '../../../test-utils'
import { Color } from 'interfaces/theme'
import { COLORS } from 'lib/constants'
import Badge from '.'

test('should render correctly', () => {
  const tree = renderer.create(<Badge variant="dot">B</Badge>).toJSON()
  expect(tree).toMatchSnapshot()
})

test('should render content', () => {
  const { getByText } = render(<Badge content={1}>B</Badge>)

  expect(getByText('1')).toBeInTheDocument()
})

test('should work with different colors', () => {
  const tree = renderer
    .create(
      <>
        {COLORS.map((color: Color) => (
          <Badge color={color} variant="dot" key={color}>
            {color.substring(0, 2)}
          </Badge>
        ))}
      </>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

test('should not display a number higher than `max`', () => {
  const { getByText } = render(
    <Badge content={100} max={99}>
      B
    </Badge>
  )

  expect(getByText('99+')).toBeInTheDocument()
})

test('should not display content when number 0', () => {
  const { getByText } = render(<Badge content={0}>B</Badge>)

  expect(getByText('0')).not.toBeVisible()
})

test('should display content when `showZero` is truthy', () => {
  const { getByText } = render(
    <Badge content={0} showZero>
      B
    </Badge>
  )

  expect(getByText('0')).toBeInTheDocument()
})
