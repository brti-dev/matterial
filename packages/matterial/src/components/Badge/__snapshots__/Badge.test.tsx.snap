// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<span
  className="container"
>
  B
  <span
    className="badge content variant--contained color--default no-hover variant--dot size--medium "
    hidden={false}
  />
</span>
`;

exports[`should work with different colors 1`] = `
Array [
  <span
    className="container"
  >
    pr
    <span
      className="badge content variant--contained color--primary no-hover variant--dot size--medium "
      hidden={false}
    />
  </span>,
  <span
    className="container"
  >
    se
    <span
      className="badge content variant--contained color--secondary no-hover variant--dot size--medium "
      hidden={false}
    />
  </span>,
  <span
    className="container"
  >
    er
    <span
      className="badge content variant--contained color--error no-hover variant--dot size--medium "
      hidden={false}
    />
  </span>,
  <span
    className="container"
  >
    wa
    <span
      className="badge content variant--contained color--warning no-hover variant--dot size--medium "
      hidden={false}
    />
  </span>,
  <span
    className="container"
  >
    in
    <span
      className="badge content variant--contained color--info no-hover variant--dot size--medium "
      hidden={false}
    />
  </span>,
  <span
    className="container"
  >
    su
    <span
      className="badge content variant--contained color--success no-hover variant--dot size--medium "
      hidden={false}
    />
  </span>,
  <span
    className="container"
  >
    da
    <span
      className="badge content variant--contained color--dark no-hover variant--dot size--medium "
      hidden={false}
    />
  </span>,
  <span
    className="container"
  >
    li
    <span
      className="badge content variant--contained color--light no-hover variant--dot size--medium "
      hidden={false}
    />
  </span>,
]
`;
